classes/app/helper/times/day.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace assignsubmission_pxaiwriter\app\helper\times;
+
+
+use DateTime;
+
+/* @codeCoverageIgnoreStart */
+defined('MOODLE_INTERNAL') || die();
+/* @codeCoverageIgnoreEnd */
+
+class day implements interfaces\day
+{
+    private DateTime $date;
+
+    public function __construct(DateTime $date)
+    {
+        $this->date = $date;
+    }
+
+    public function get_date_time(): DateTime
+    {
+        return $this->date;
+    }
+
+    public function get_start_of_day(): DateTime
+    {
+        $start_of_day = clone $this->date;
+        $start_of_day->setTime(0, 0, 0);
+        return $start_of_day;
+    }
+
+    public function get_end_of_day(): DateTime
+    {
+        $start_of_day = clone $this->date;
```

Comment:
Just out of curiosity: why use clone instead of just referencing $this->date? I don't think I've ever seen that in use before :smile:

Reply:
Because ``$this->date`` is not an immutable ``DateTime`` object and we want a copy of object not the reference of it.
And because of that if you modified something that return by ``get_end_of_day()``
It will modify the ``$this->date`` itself.  Which it is undesire behavior.

You can also doing ``new DateTime($this->date->format('Y-m-d H:i:s'), $this->date->getTimezone())``.

And in this case ``clone`` is a bit easier to read and it copy the object without an extra work :smile: 

There is also ``DateTimeImmutable`` that can also be use in this situation.

Example
```php


$date = new DateTime('2023-02-01 00:00:00');
$a = $date;
$b = $date;
$a->setTime(12, 0, 0);
$b->setTime(9, 0, 0);

echo $a->format('Y-m-d H:i:s'); // 2023-02-01 09:00:00
echo PHP_EOL;
echo $b->format('Y-m-d H:i:s'); // 2023-02-01 09:00:00

$date = new DateTime('2023-02-01 00:00:00');
$a = clone $date;
$b = clone $date;

$a->setTime(12, 0, 0);
$b->setTime(9, 0, 0);

echo PHP_EOL;
echo $a->format('Y-m-d H:i:s'); // 2023-02-01 12:00:00
echo PHP_EOL;
echo $b->format('Y-m-d H:i:s'); // 2023-02-01 09:00:00

```